src/ser.rs

```diff
@@ -0,0 +1,183 @@
+pub trait Deserialization {
+    /// Constructs a graph from tuple data (eclass ids as `Id`).
+    fn from_tuples<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<Id>)>) -> (Self, ColorPalette) where Self: Sized;
+    /// Constructs a graph from tuple data (eclass ids as numbers).
+    fn from_tuples_int<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<usize>)>) -> (Self, ColorPalette) where Self: Sized;
+    /// Parses an constructs a graph from textual representation.
+    fn from_tuples_text(in_: &mut impl Read) -> Result<(Self, ColorPalette)> where Self: Sized;
+}
+
+trait DeserializationHelper<L: Language> {
+    fn add_class<'a>(&'a mut self, id: Id);
+    fn add_node(&mut self, eclass: Id, enode: L) -> &L;
+    fn update_parents(&mut self, parent: Id, enode: &L);
+    fn line_to_tuple(line: &str) -> (String, Vec<Id>);
+}
+
+pub struct ColorPalette {
+    colors: HashMap<Id, ColorId>
+}
+
+impl ColorPalette {
+    fn new() -> Self {
+        ColorPalette { colors: HashMap::default() }
+    }
+
+    fn get<L: Language, N: Analysis<L>>(&mut self, g: &mut EGraph<L, N>, color: Id) -> ColorId {
+        match self.colors.get(&color) {
+           Some(cid) => *cid,
+           None => {
+               let v = g.create_color();
+               self.colors.insert(color, v);
+               v
+           }
+        }
+    }
+}
+
+impl Default for ColorPalette {
+    fn default() -> Self { ColorPalette::new() }
+}
+
+impl<N: Analysis<SymbolLang>> Serialization for EGraph<SymbolLang, N> {
+    fn to_tuples(&self) -> Vec<(String, Vec<Id>)> {
+        self.classes().flat_map(
+            |ec| {
+                let target = ec.id;
+                ec.nodes.iter().map(
+                    move |u| (u.op.to_string(), [vec![target], u.children.clone()].concat()))
+            }).collect::<Vec<_>>()
+    }
+
+    fn to_tuples_text(&self, palette: &ColorPalette, out: &mut impl Write) -> Result<()>{
+        // Write edges
+        for (op, ids) in self.to_tuples() {
+            writeln!(out, "{op} {ids}", ids = ids.iter().join(" "))?;
+        }
+        // Write colors
+        let id_start = self.classes().map(|e| e.id).max().unwrap();
+        for (i, color) in self.colors().enumerate() {
+            let color_id =
+                if let Some(color_entry) = palette.colors.iter().find(|e| e.1 == &color.get_id()) {
+                    usize::from(*color_entry.0)
+                }
+                else {
+                    let color_id = usize::from(id_start) + i + 1;
+                    writeln!(out, "clr#{i} {color_id}")?;
+                    color_id
+                };
+            for id in color.black_reps() {
+                writeln!(out, "?~ {color_id} {members}",
+                         members = color.black_ids(*id).unwrap().iter().join(" "))?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<L: Language, N: Analysis<L>> DeserializationHelper<L> for EGraph<L, N> {
```

Comment:
Do we really need desiaralization? We are missing some of the inner fields of the EGraph like this